From 2f1573fe27de4008faedd38ee9cdadc3eec31909 Mon Sep 17 00:00:00 2001
From: Viktor Malik <viktor.malik@gmail.com>
Date: Fri, 6 Jun 2025 07:54:19 +0200
Subject: [PATCH] runqlen.bt: Use old version of the tool

RHEL 9 doesn't have 736c55a02c47 ("sched/fair: Rename cfs_rq.nr_running
into nr_queued") so we need to use the old name of the struct fields.

Signed-off-by: Viktor Malik <viktor.malik@gmail.com>
---
 tools/old/runqlen.bt | 41 -----------------------------------------
 tools/runqlen.bt     |  8 +++++---
 2 files changed, 5 insertions(+), 44 deletions(-)
 delete mode 100755 tools/old/runqlen.bt

diff --git a/tools/old/runqlen.bt b/tools/old/runqlen.bt
deleted file mode 100755
index 23a7fb85..00000000
--- a/tools/old/runqlen.bt
+++ /dev/null
@@ -1,41 +0,0 @@
-#!/usr/bin/env bpftrace
-/*
- * runqlen.bt	CPU scheduler run queue length as a histogram.
- *		For Linux, uses bpftrace, eBPF.
- *
- * This is a bpftrace version of the bcc tool of the same name.
- *
- * For Linux < 6.14.
- *
- * Copyright 2018 Netflix, Inc.
- * Licensed under the Apache License, Version 2.0 (the "License")
- *
- * 07-Oct-2018	Brendan Gregg	Created this.
- */
-
-#ifndef BPFTRACE_HAVE_BTF
-#include <linux/sched.h>
-
-// Without BTF, we'll need to declare some of this struct manually,
-// since it isn't available to be #included. This will need maintenance to match
-// your kernel version. It is from kernel/sched/sched.h:
-struct cfs_rq {
-	struct load_weight load;
-	unsigned int nr_running;
-	unsigned int h_nr_running;
-};
-#endif
-
-BEGIN
-{
-	printf("Sampling run queue length at 99 Hertz... Hit Ctrl-C to end.\n");
-}
-
-profile:hz:99
-{
-	$task = (struct task_struct *)curtask;
-	$my_q = (struct cfs_rq *)$task->se.cfs_rq;
-	$len = (uint64)$my_q->nr_running;
-	$len = $len > 0 ? $len - 1 : 0;	// subtract currently running task
-	@runqlen = lhist($len, 0, 100, 1);
-}
diff --git a/tools/runqlen.bt b/tools/runqlen.bt
index 394a7b3d..23a7fb85 100755
--- a/tools/runqlen.bt
+++ b/tools/runqlen.bt
@@ -5,6 +5,8 @@
  *
  * This is a bpftrace version of the bcc tool of the same name.
  *
+ * For Linux < 6.14.
+ *
  * Copyright 2018 Netflix, Inc.
  * Licensed under the Apache License, Version 2.0 (the "License")
  *
@@ -19,8 +21,8 @@
 // your kernel version. It is from kernel/sched/sched.h:
 struct cfs_rq {
 	struct load_weight load;
-	unsigned int nr_queued;
-	unsigned int h_nr_queued;
+	unsigned int nr_running;
+	unsigned int h_nr_running;
 };
 #endif
 
@@ -33,7 +35,7 @@ profile:hz:99
 {
 	$task = (struct task_struct *)curtask;
 	$my_q = (struct cfs_rq *)$task->se.cfs_rq;
-	$len = (uint64)$my_q->nr_queued;
+	$len = (uint64)$my_q->nr_running;
 	$len = $len > 0 ? $len - 1 : 0;	// subtract currently running task
 	@runqlen = lhist($len, 0, 100, 1);
 }
-- 
2.49.0

